### PR TITLE
allow custom credentials through environment variables

### PR DESCRIPTION
**Pull request by mlintern**

### CODE REVIEW

File: lib/models/account.js
Comment (fishcharlie) [inline lib/models/account.js:25]: It looks like this is being used in 2 different places. Maybe we should move these variables out into a separate file and have a single location to make the code more DRY?

OR have the version in `lib/controllers/bucket.js` reference this one.

**Review by fishcharlie — APPROVED** [review]
One suggestion. Not sure it should hold up the PR tho.

File: lib/controllers/bucket.js
Comment (leontastic) [inline lib/controllers/bucket.js:32]: I think this can simply be a dummy string as long as it matches the reference above on line 15, since the token is not meant to be decoded by anyone but the S3 server. That would eliminate the reference to `secretAccessKey` here.

**Review by leontastic — COMMENTED** [review]
This is the right approach.

If we're going to support configurable credentials we should probably initialize them in the `S3rver` constructor instead. That way we can support `accessKeyId` and `secretAccessKey` directly in the CLI and API without referencing `process.env`, and leave it up to the user how they want to load their keys (environment variables, via CLI, from file, etc.):

```js
new S3rver({
  // other config options
  defaultAccessKeyId: 'MY_ACCESS_KEY_ID',
  defaultSecretAccessKey: 'MY_SECRET_ACCESS_KEY'
})
```

```bash
$ s3rver --directory s3temp --defaultAccessKeyId "MY_ACCESS_KEY_ID" --defaultSecretAccessKey "MY_SECRET_ACCESS_KEY"
```

File: lib/controllers/bucket.js
Comment (kherock) [reply lib/controllers/bucket.js:32]: Yeah this function is definitely missing some documentation explaining what's going on, my bad.

Basically, there's evidence that S3 uses a symmetric key to generate continuation tokens since the token length is proportional to the combined region, bucket name, and object key. The string I used here is arbitrary. If you'd like to have something more descriptive there instead, `'S3RVER_CTOKEN_KEY'` would probably be fine.

**Comment by kherock** [comment]
@leontastic would it be too magical if we inferred default values for those based on credentials loaded in the environment (i.e. ask `aws-sdk` for its currently loaded credentials)?

I did intentionally leave out configurability in the constructor in the original dummy account implementation because to me an S3 API with the concept of "default account" felt like technical debt for adding proper bucket ownership later. Cross-account scenarios are fairly common, but I'm also aware in that case that you're probably a large company who has the resources to easily run those tests against the real thing anyway.

**Comment by leontastic** [comment]
@kherock I don't think it's too magical, but it would force users of the JavaScript API to use multiple configuration methods. Imagine using `dotenv` to load environment variables and instantiating your S3rver instance with JS:

```js
require('dotenv').config()

new S3rver({
  directory: process.env.TMP_S3_DIRECTORY,
  // ... more config
})
```

With the proposed change the only way you could add a default key pair is either baking it into `process.env` (configuration by side-effect):

```js
require('dotenv').config()

process.env.S3RVER_ACCESS_KEY_ID = process.env.AWS_ACCESS_KEY_ID

new S3rver({
  directory: process.env.TMP_S3_DIRECTORY,
  // ... more config
})
```

or repeating the key in your `.env` file, or defining the environment variable on the command line, in addition to the config you wrote in your setup script. Compare this to supporting the config options out of the box:

```js
require('dotenv').config()

new S3rver({
  directory: process.env.TMP_S3_DIRECTORY,
  defaultAccessKeyId: process.env.AWS_ACCESS_KEY_ID,
  // ... more config
})
```

The way that the `bin` file is written means that all the default constructor options can be supported by the CLI without any extra code, we just have to document the options.

I spent a few hours today writing a PR inspired by this one, but which accomplishes this configuration via constructor. 
Thankfully the concept of accounts is very thin and sparsely used in the project so it wasn't too difficult. I think it would be a really nice convenience to be able to configure a default user and keypair.

> I did intentionally leave out configurability in the constructor in the original dummy account implementation because to me an S3 API with the concept of a "default account" felt like technical debt for adding proper bucket ownership later. Cross-account scenarios are fairly common, but I'm also aware in that case that you're probably a large company who has the resources to easily run those tests against the real thing anyway.

I agree with you this would incur technical debt because any internal abstraction would deviate from how AWS actually handles accounts, authentication, and ownership. IMO there's no way we could faithfully model the AWS authentication model perfectly (we'd basically be building mock IAM at that point) but it is probably sufficient and useful to users of S3rver if we provide a thin abstraction that allows them to bootstrap default accounts and even add some more. It wouldn't behave perfectly like S3, but if there is demand for it, it could form the basis for work on mocking cross-account or multiple keypair use cases in the future.

**Comment by kherock** [comment]
@leontastic that's all valid, thanks for taking the time to address this! I just want to clarify that what I'm really after is a zero-config solution that works for most people. If the defaults work for most people, that's fewer people to annoy when we come up with something that messes with people using more advanced configuration. Something along the lines of
```javascript
const { accessKeyId, secretAccessKey } = require('aws-sdk').config.credentials;
defaultAccessKeyId ??= accessKeyId;
defaultSecretAccessKey ??= secretAccessKey;
```
in the constructor would solve cases like #498 without introducing new config. Constructor options are definitely the correct way to set this up, and I'd like for their default values to satisfy most cases.

**Comment by leontastic** [comment]
@kherock That is a very good point, I will see if I can incorporate values from `aws-sdk` into the defaults in the refactor. I like the idea of bothering users less for their config if it's not important to them, and letting them override if they want to test something specific (e.g. auth with signing tokens https://github.com/jamhall/s3rver/issues/659 https://github.com/jamhall/s3rver/issues/534).

**Pull request closed without merging**
